Tellor grant application

brendaloya opened the pull request:
* added Tellor's application

* chore: clarify proposal

* updated reporter repo

* updated tellor.md

---------

Co-authored-by: Frank Bell <frank@parity.io>

#### Project Abstract

Oracles are a key piece of infrastructure for any blockchain ecosystem. Tellor proposes a crypto economically incentivized oracle that is completely flexible for the needs of any data type or frequency.  For this Tellor will implement an oracle pallet along with a deployment of controller smart contracts on a to-be-selected EVM parachain that will handle the reporter staking staking, governance (for disputing data validity), and reporter reward system. This will enable parachains across the ecosystem to read, and/or request a push of oracle data. Once the network of reporters is bootstrapped, there is minimal team intervention needed. Anyone can become a reporter by staking, anyone can dispute the validity of the data for a fee, and the user can incentivize reporters to quickly add or replace the data they need by posting a reward. The system is permissionless, as well as sybil and censorship resistant. 

#### Grant [level](https://github.com/w3f/Grants-Program#level_slider-levels)

- [ ] **Level 1**:  Up to $10,000, 2 approvals
- [ ] **Level 2**:  Up to $30,000, 3 approvals
- [x] **Level 3**:  Unlimited, 5 approvals (for >$100k: Web3 Foundation Council approval)

#### Application Checklist

- [x] The [application template](https://github.com/w3f/Grants-Program/blob/master/applications/application-template.md) has been copied and aptly renamed (`project_name.md`).
- [x] I have read the [application guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/grant_guidelines_per_category.md).
- [x] A BTC, Ethereum (USDC/DAI) or Polkadot/Kusama (aUSD/USDT) address for the payment of the milestones is provided in the application.
- [x] The software delivered for this grant will be released under an open-source license specified in the application.
- [x] The initial PR contains only one commit (squash and force-push if needed).
- [x] The grant will only be announced once the first milestone [has been accepted](https://github.com/w3f/Grant-Milestone-Delivery#process) (see the [announcement guidelines](https://github.com/w3f/Grants-Program/blob/master/docs/Support%20Docs/announcement-guidelines.md)).
- [ ] I prefer the discussion of this application to take place in a private Element/Matrix channel. My username is: `@_______:matrix.org` (change the homeserver if you use a different one)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/w3f/Grants-Program?pullRequest=1532) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/w3f/Grants-Program?pullRequest=1532) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/w3f/Grants-Program?pullRequest=1532) it.</sub>

keeganquigley commented:
Thanks for the interesting application @brendaloya. As far as I'm aware, neither the Chainlink [oracle pallet](https://github.com/smartcontractkit/chainlink-polkadot) nor the Chainlink [price feed pallet](https://github.com/smartcontractkit/chainlink-polkadot/tree/master/pallet-chainlink-feed) are maintained anymore, which makes your application intriguing. Couple of questions/comments:

- Please note that we only fund the development of software that doesn't exist yet. Since you mentioned that most of the EVM smart contracts have already been built, perhaps you could remove that part from the milestones and reduce the price/scope accordingly. If you reduced it to 30k it would become a level 2 grant, and would only need 3 approvals from the committee.
- Each milestone should have its own table that breaks down the work into smaller concrete deliverables that can be evaluated/tested. The way you have it formatted currently would equate to 2 milestones. Could you update it to either remove the extra milestones and stick with 2, or add tables for each of them?
- Since you are a new team for us, you may want to reduce the scope to 2 or 3 milestones, and then apply for a follow-up grant after the committee has had a chance to see some of your work. The application might have better chances of approval this way.
- Have you heard of the [Acurast pallet](https://docs.acurast.com/integrations/substrate/)? I wonder how similar your project is, and whether [they](https://acurast.com/) could be useful for you in helping define off-chain data and computation requirements for substrate.
- Not that you mentioned one, but in any case please take note of our [policy](https://github.com/w3f/Grants-Program#guidelines) when it comes to token sales.
- Has the reporter incentivisation scheme been working successfully on other chains?

brendaloya commented:
Thanks for the feedback @keeganquigley. I've added comments below and made some updates changed the application to two milestones. 

Thanks for the interesting application @brendaloya. As far as I'm aware, neither the Chainlink [oracle pallet](https://github.com/smartcontractkit/chainlink-polkadot) nor the Chainlink [price feed pallet](https://github.com/smartcontractkit/chainlink-polkadot/tree/master/pallet-chainlink-feed) are maintained anymore, which makes your application intriguing. Couple of questions/comments:

-Response: Once implemented our proposed solution could function with zero team interaction because anyone can become a reporter as long as they stake,  reporters are incentivized by consumers to provide data through payments, the data validation is open to anyone that pays the fee to dispute and everyone is incentivized to dispute because you get the reporter stake if the dispute is successful otherwise the reporter is compensated with the fee for reward/consumer payment losses. I can't speak for another project but truly decentralized and crypto-economically incentivized systems do not stop working because the team does not maintain them. However, we do plan to help with the reporter bootstrapping and future projects' integrations as demand and usage of Parity/Polkadot ecosystem increases. 

Please note that we only fund the development of software that doesn't exist yet. Since you mentioned that most of the EVM smart contracts have already been built, perhaps you could remove that part from the milestones and reduce the price/scope accordingly. If you reduced it to 30k it would become a level 2 grant, and would only need 3 approvals from the committee. 

-Response: Some of the EVM contracts exist, however, some have to be developed and there will be some updates to ensure these work with the Parity/Polkadot ecosystem and tested.

Each milestone should have its own table that breaks down the work into smaller concrete deliverables that can be evaluated/tested. The way you have it formatted currently would equate to 2 milestones. Could you update it to either remove the extra milestones and stick with 2, or add tables for each of them?  
-Response: I changed these to be two milestones now. 

Since you are a new team for us, you may want to reduce the scope to 2 or 3 milestones, and then apply for a follow-up grant after the committee has had a chance to see some of your work. The application might have better chances of approval this way. 
-Response: If this is approved we will do a follow-up grant for mainnet. Tellor has been live since August 2019, all of our code is open source https://github.com/tellor-io/ and happy to walk the committee through any of our work.

Have you heard of the Acurast pallet? I wonder how similar your project is, and whether they could be useful for you in helping define off-chain data and computation requirements for substrate. 

-Response: We'll look into it but at first glance I can't easily find how becoming a data transmitter is open to anyone or if that how hat is decentralized. They also don't seem to have any penalties for these providing bad data or a live dispute process. Our system is crypto-economically secured and has a dispute process in place. 

Not that you mentioned one, but in any case please take note of our policy when it comes to token sales. 
-Response: Will do. However, we don't plan to launch a new token and TRB is currently available at Coinbase and Binance. Our proposed solution would rely on bridged TRB or the consumer chain's token. 

Has the reporter incentivization scheme been working successfully on other chains? 
-Response: Yes, since August 2019.

keeganquigley reviewed (dismissed):
thanks for the changes @brendaloya I have marked your application as ready for review and will share it with the rest of the committee.

Seeing as this is an interesting XCM use case, and we need more work done on oracles, I'm personally happy to support it.

dsm-w3f reviewed (commented):
Thank you for your application. It is a very interesting one. I notice that in milestone 1, deliverable 3, there is no specification in this item. Could you please add it? Please notice that we usually perform system tests on deliverables as part of the Milestone evaluation process which requires a tutorial for manual testing or automated e2e tests.

Noc2 reviewed (changes requested):
Thanks a lot for the application. I have a couple of questions/comments:

- Currently, the headline for milestone 2 is missing. Could you please add this? 
- Your first milestone focuses on solidity smart contracts. We usually prefer to support the development of ink! Smart contracts or other solutions that are native to our ecosystem. Would you be willing to integrate this (at least initial PoCs) into the application? I saw that you mentioned that you plan on working on WASM smart contracts. 
- Could you remove the Tellor Tributes (TRB) Token from the application? See our guidelines: https://github.com/w3f/Grants-Program#guidelines 
- Could you also add a reference to the definition that is listed above in the document for the integration contract ("with the functionality described above")? In general, feel free to add more details to the milestone tables also for the pallet since these are the requirements of our contracts. 
- Finally, do you have plans to integrate [off-chain works](https://docs.substrate.io/fundamentals/offchain-operations/) into your oracle solution?

semuelle commented on applications/Tellor.md:
```diff
@@ -0,0 +1,231 @@
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1 | Develop Controller contracts | We will provide a set of solidity smart contracts with the functionality described above  |
+| 2 | Develop Parachain integration contract| We will provide an integration contract |
+| 3 | Local testing| |
```
As @dsm-w3f said, this still requires a specification.

brendaloya replied:
You are correct deliverable 0c should cover the testing so I’ve removed deliverable 3 from milestone 1.

semuelle commented on applications/Tellor.md:
```diff
@@ -0,0 +1,231 @@
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1 | Substrate Oracle pallet design and integration | We will provide the Substrate oracle pallet                                                                                                                     |
+| 2 | Testing the pallet and testing integration with projects and selected parachains| We will test full functionality of the system for interactions between the consumer chain and oracle pallet                                                     |
```
What exactly is being delivered here? A test report? There is already a `Testing and Testing Guide` deliverable above, how is this different?

brendaloya replied:
Hi @semuelle ,
I think it could be part of deliverable *0c but I just wanted to spell out that we have to test milestone 1 and 2 interoperability, happy to remove it though.

semuelle reviewed (changes requested):
@dsm-w3f and @Noc2 already pointed out the issues I found. Other than that, this is a good proposal and I'd be happy to support it.

semuelle commented:
pinging @brendaloya

brendaloya commented:
> thanks for the changes @brendaloya I have marked your application as ready for review and will share it with the rest of the committee.
> 
> Seeing as this is an interesting XCM use case, and we need more work done on oracles, I'm personally happy to support it.

Response: 
Thank you for the feedback  @dsm-w3f. You are correct deliverable 0c should cover the testing so I’ve removed deliverable 3 from milestone 1.

brendaloya commented:
Hi @Noc2 ,

> Thanks a lot for the application. I have a couple of questions/comments:
> 
> * Currently, the headline for milestone 2 is missing. Could you please add this?
-Response: Done 
> * Your first milestone focuses on solidity smart contracts. We usually prefer to support the development of ink! Smart contracts or other solutions that are native to our ecosystem. Would you be willing to integrate this (at least initial PoCs) into the application? I saw that you mentioned that you plan on working on WASM smart contracts.
-Response: We would love to explore it but I currently don’t have anyone in the team that is fluent in ink! and the proposal right now tries to minimize costs. Currently we plan on creating contracts enable interaction with the XCM and making changes to fit Parity’s parachains ecosystem. Rewriting the existing contracts on a different language and getting an audit on that language would make this iteration very expensive. We would love to see some usage and if users begin demanding the change to ink! for integrations, we would be happy to explore it. 
> * Could you remove the Tellor Tributes (TRB) Token from the application? See our guidelines: https://github.com/w3f/Grants-Program#guidelines
-Response: Done
> * Could you also add a reference to the definition that is listed above in the document for the integration contract ("with the functionality described above")? In general, feel free to add more details to the milestone tables also for the pallet since these are the requirements of our contracts.
-Response: Added the reference and some clarification but let me know if more details are needed. We will have a full spec. 
> * Finally, do you have plans to integrate [off-chain works](https://docs.substrate.io/fundamentals/offchain-operations/) into your oracle solution?
-Response: not currently but happy to explore and brainstorm if it would be a good fit.

semuelle commented on applications/Tellor.md:
```diff
@@ -0,0 +1,228 @@
+| **0c.** | Testing and Testing Guide | Core functions will be fully covered by comprehensive unit tests to ensure functionality and robustness. In the guide, we will describe how to run these tests. |
+| **0d.** | Docker | We will provide a Dockerfile(s) that can be used to test all the functionality delivered with this milestone. |
+| 1 | Substrate Oracle pallet design and integration | We will provide the Substrate oracle pallet |
+| 2 | Testing functionallity of the the pallet and testing integration with a mock project and selected parachains| We will test full functionality of the system for interactions between the consumer chain and oracle pallet (meaning test the functinallity between milestone 1 and 2 -  solidity contracts, pallet, XCM)|
```
Please specify what is being delivered here. A testing guide? Test report? Article?

brendaloya replied:
Hi @semuelle , thanks for the feedback.

> Please specify what is being delivered here. A testing guide? Test report? Article?

-Response: On deliverable 2, what is being delivered is just more testing. I read deliverable 0c as being more specific to the pallet testing (deliverable 1) so in deliverable 2 it would just be more test delivered/included that test the interoperability between milestone 1 and milestone 2 deliverable 1. I could remove it if you think that is covered by deliverable 0c.

semuelle replied:
No, that's fine given the architecture. It just needs to be clear from the specification of the deliverable what exactly will be delivered, so that we know what to expect. You could add, for example, "we will provide a tutorial how to replicate these tests locally."

brendaloya replied:
Hi @semuelle , I've added clarification on milestone 2 deliverable 1 since we'll definitely provide a guide on running the additional tests.

brendaloya commented:
> Just for transparency: I have updated your [comment above](https://github.com/w3f/Grants-Program/pull/1532#issuecomment-1444152236) for readability. Github was interpreting everything as one large quote.

Thanks @semuelle

semuelle reviewed (approved):
LGTM.

takahser reviewed (approved):
This sounds interesting, especially since the Chainlink Substrate oracle pallet hasn't been updated in a while and doesn't seem to be actively maintained anymore. Happy to approve.

keeganquigley commented:
Hi @brendaloya how is milestone 2 coming along?

keeganquigley commented:
pinging @brendaloya

keeganquigley commented:
Hi @brendaloya are you still working on this grant? Please note that if we don't hear back from you after two weeks we will file a PR to close the grant due to inactivity. Thank you for understanding.
